pkg/addonmanager/cloudevents/manager.go

```diff
@@ -0,0 +1,176 @@
+	config := a.GetConfig()
+	addonAgents := a.GetAddonAgents()
+
+	var addonNames []string
+	for key := range addonAgents {
+		addonNames = append(addonNames, key)
+	}
+
+	// To support sending ManifestWorks to different drivers (like the Kubernetes apiserver or MQTT broker), we build
+	// ManifestWork client that implements the ManifestWorkInterface and ManifestWork informer based on different
+	// driver configuration.
+	// Refer to Event Based Manifestwork proposal in enhancements repo to get more details.
+	_, clientConfig, err := cloudeventswork.NewConfigLoader(a.options.WorkDriver, a.options.WorkDriverConfig).
+		WithKubeConfig(a.GetConfig()).
+		LoadConfig()
+	if err != nil {
+		return err
+	}
+
+	// we need a separated filtered manifestwork informers so we only watch the manifestworks that manifestworkreplicaset cares.
+	// This could reduce a lot of memory consumptions
+	workInformOption := workinformers.WithTweakListOptions(
+		func(listOptions *metav1.ListOptions) {
+			selector := &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{
+						Key:      addonv1alpha1.AddonLabelKey,
+						Operator: metav1.LabelSelectorOpIn,
+						Values:   addonNames,
+					},
+				},
+			}
+			listOptions.LabelSelector = metav1.FormatLabelSelector(selector)
+		},
+	)
+
+	clientHolder, err := cloudeventswork.NewClientHolderBuilder(clientConfig).
+		WithClientID(a.options.CloudEventsClientID).
+		WithSourceID(a.options.SourceID).
+		WithInformerConfig(10*time.Minute, workInformOption).
+		WithCodecs(codec.NewManifestBundleCodec()).
+		NewSourceClientHolder(ctx)
+	if err != nil {
+		return err
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	addonClient, err := addonv1alpha1client.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	clusterClient, err := clusterv1client.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	addonInformers := addoninformers.NewSharedInformerFactory(addonClient, 10*time.Minute)
+	clusterInformers := clusterv1informers.NewSharedInformerFactory(clusterClient, 10*time.Minute)
+	dynamicInformers := dynamicinformer.NewDynamicSharedInformerFactory(dynamicClient, 10*time.Minute)
+
+	kubeInformers := kubeinformers.NewSharedInformerFactoryWithOptions(kubeClient, 10*time.Minute,
+		kubeinformers.WithTweakListOptions(func(listOptions *metav1.ListOptions) {
+			selector := &metav1.LabelSelector{
+				MatchExpressions: []metav1.LabelSelectorRequirement{
+					{
+						Key:      addonv1alpha1.AddonLabelKey,
+						Operator: metav1.LabelSelectorOpIn,
+						Values:   addonNames,
+					},
+				},
+			}
+			listOptions.LabelSelector = metav1.FormatLabelSelector(selector)
+		}),
+	)
+
+	workClient := clientHolder.WorkInterface()
+	workInformers := clientHolder.ManifestWorkInformer()
+
+	// addonDeployController
+	err = workInformers.Informer().AddIndexers(
```

Comment:
nit: the indexer part should be put in a shared dir. It is duplicated with the manager.